Sources/SwiftFusion/Geometry/Pose2.swift

```diff
@@ -1,34 +1,223 @@
+  @derivative(of: init(_:_:))
+  @usableFromInline
+  static func vjpInit(_ t: Point2, _ r: Rot2)
+    -> (value: Pose2, pullback: (TangentVector) -> (Point2.TangentVector, Rot2.TangentVector))
+  {
+    // Explanation of this calculation:
+    //
+    // We would like the differential of `init` at `(t, r)`, `dinit_(t, r)`, to satisfy:
+    //   init(t * exp(eps_t), r * exp(eps_r)) ~= init(t, r) * exp(dinit_(t, r)(eps_t, eps_r))
+    // where `eps_t`, `eps_r` are "small" elements of R^2 and so(2) respectively.
+    //
+    // Multiplying on the left by `init(t, r)^-1` gives
+    //   exp(dinit_(t, r)(eps_t, eps_r))
+    //     = init(t, r)^-1 * init(t * exp(eps_t), r * exp(eps_r))
+    //     = init(r^-1 * (-t), inverse(r)) * init(t * exp(eps_t), r * exp(eps_r))
+    //     = init(r^-1 * exp(eps_t), exp(eps_r))
+    //
+    // Since `eps_t`, `eps_r` are small, we can take logs of both sides to get
+    //   dinit_(t, r)(eps_t, eps_r) = (r^-1 * eps_t, eps_r)
+    //
+    // Or in matrix notation:
+    //   dinit_(t, r)(eps_t, eps_r) =  ( r^-1  0 ) ( eps_t )
+    //                                 ( 0     1 ) ( eps_r )
+    //
+    // We actually need the pullback rather than the differential, so that's the transpose:
+    //   pbinit_(t, r)(eps_t, eps_r) =  ( r 0 ) ( eps_t )
+    //                                  ( 0 1 ) ( eps_r )
+    //
+    // The pullback here implements the linear map corresponding to that matrix.
+    (Pose2(t, r), { eps in
+      (
+        Point2.TangentVector(
+          x: r.c * eps.vx - r.s * eps.vy,
+          y: r.s * eps.vx + r.c * eps.vy
+        ),
+        eps.omega
+      )
+    })
+  }
+
+  /// The derivative of `t` satisfiying convention (2) described above.
+  @derivative(of: t)
+  @usableFromInline
+  func vjpT() -> (value: Point2, pullback: (Point2.TangentVector) -> TangentVector) {
+    // Explanation of this calculation:
+    //
+    // `t` is the inverse of `init`, followed by a projection to the translation component. So we
+    // can get the matrix for the pullback of `t` by inverting the matrix for the pullback of
+    // `init` and then taking the relevant columns.
+    //
+    // The inverse is:
+    //   ( r^-1 0 )
+    //   ( 0    1 )
+    //
+    // The below implements the linear transformation represented by the first column of this matrix.
+    return (t, { v in
+      TangentVector(
+        vx: rot.c * v.x + rot.s * v.y,
+        vy: -rot.s * v.x + rot.c * v.y,
+        omega: .zero
+      )
+    })
+  }
+
+  /// The derivative of `t` satisfiying convention (2) described above.
+  @derivative(of: rot)
+  @usableFromInline
+  func vjpRot() -> (value: Rot2, pullback: (Rot2.TangentVector) -> TangentVector) {
+    // This pullback is similar to the pullback in `vjpT`, but for the second column of the
+    // matrix. See the comment in `vjpT` for more information.
+    return (rot, { TangentVector(vx: 0, vy: 0, omega: $0) })
+  }
+
+  /// The Lie algebra se(2).
+  public struct TangentVector:
```

Comment:
Yeah, it's beautiful to have these, but again: Vector3 will do fine for our purposes, and gives us vector space properties for free. (and they are easy to concatenate them later into TangentVectors for product manifolds)

Reply:
ah yes, "easy to concatenate them later into TangentVectors for product manifolds" is the thing that makes it easy to linearize factor graphs